packages/belos/src/BelosSolverFactory.hpp

```diff
@@ -74,7 +58,46 @@
 #include <stdexcept>
 #include <vector>
 
+// For convenience have a specialized form of registerSolverFactoryForLib()
+// called at construction of SolverFactoryParent which will pick up the proper
+// lib (belos, belosepetra, or belosepetra) based on the template parameters.
+// Then registerSolverFactoryForLib() will link to and register all the managers
+// for the proper lib. These includes were added for the specialization.
+
+// for belos lib
+#include "BelosMultiVec.hpp"
+#include "BelosOperator.hpp"
+
+// for belosepetra lib
+#ifdef HAVE_BELOS_EPETRA
+#include "Epetra_MultiVector.h"
+#include "Epetra_Operator.h"
+#endif
+
+// for belostpetra lib
+#ifdef HAVE_BELOS_TPETRA
+#include "Tpetra_MultiVector.hpp"
+#include "Tpetra_Operator.hpp"
+#include "TpetraCore_ETIHelperMacros.h"
+#endif
+
 namespace Belos {
+
+// Declare the registration methods for the 3 different libs
```

Comment:
This is one place where weak symbols could be useful; it would let us put the definitions (and even declarations) of these functions downstream.  In general, I'm a bit uncomfortable at needing to put these functions in the main Belos "subpackage" as opposed to downstream, but I'm not sure how else to do it.  It might still make sense to have a separate header file for each of these declarations.